tests/jballerina-unit-test/src/test/resources/test-src/record/readonly_record_fields.bal

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+const INHERENT_TYPE_VIOLATION_REASON = "{ballerina/lang.map}InherentTypeViolation";
+
+function testReadonlyRecordFields() {
+    testRecordWithSimpleReadonlyFields();
+    testInvalidRecordSimpleReadonlyFieldUpdate();
+    testValidUpdateOfPossiblyReadonlyFieldInUnion();
+    testInvalidUpdateOfPossiblyReadonlyFieldInUnion();
+}
+
+type Student record {
+    readonly string name;
+    readonly int id?;
+};
+
+function testRecordWithSimpleReadonlyFields() {
+    Student st = {
+        name: "Maryam"
+    };
+    assertEquality("Maryam", st.name);
+    assertEquality((), st?.id);
+
+    st.id = 1234; // Valid since first update.
+    assertEquality("Maryam", st.name);
+    assertEquality(1234, st?.id);
+}
+
+function testInvalidRecordSimpleReadonlyFieldUpdate() {
+    Student st1 = {
+        name: "Maryam"
+    };
+    st1.id = 1234; // Valid since first update.
+
+    // Invalid updates.
+    var fn1 = function () {
+        record {} rec = st1;
+        string str = "name";
+        rec[str] = "Jo";
+    };
+    error? res = trap fn1();
+    assertTrue(res is error);
+
+    error err = <error> res;
+    assertEquality(INHERENT_TYPE_VIOLATION_REASON, err.reason());
+    assertEquality("cannot update 'readonly' field 'name' in record of type 'Student'", err.detail()?.message);
+
+    var fn2 = function () {
+        st1.id = 4567;
+    };
+    res = trap fn2();
+    assertTrue(res is error);
+
+    err = <error> res;
+    assertEquality(INHERENT_TYPE_VIOLATION_REASON, err.reason());
+    assertEquality("cannot update 'readonly' field 'id' in record of type 'Student'", err.detail()?.message);
+}
+
+type Employee record {
+    readonly Details details;
+    string department;
+};
+
+type ReadonlyName record {
+    readonly string name;
+};
+
+type Details record {
+    string name;
+    int id;
+};
+
+function testValidUpdateOfPossiblyReadonlyFieldInUnion() {
+    Details d = {
+        name: "Jo",
+        id: 1234
+    };
+
+    Student|Details sd = d;
+    sd.name = "May";
```

Comment:
Need to clarify this behavior. `Student.name` is a readonly field, So `sd.name` assignment can panic at runtime. 
We have similar behavior on the inherent type violation. So this is OK IMO.

Reply:
Created https://github.com/ballerina-platform/ballerina-spec/issues/495 to clarify.